### ISSUE TITLE
[Feature] Move from cache based on age

### ISSUE THREAD

**Issue by tigattack** [monstermuffin/mergerfs-cache-mover#4]
Currently, cache_mover will move files from cache to the disk pool when cache usage is over the configured threshold.

This is fine, but some use cases may result in not hitting the threshold for a long time, in turn resulting in data sitting in cache for a long time.

For example, I've not written much to my storage for a few weeks, but Frigate is constantly writing a rolling ~180GB of data. This results in ~20% usage on my cache disk, so none of this is never moved to the pool.

An age threshold that acts secondarily to the utilisation threshold would be useful for such cases. This logic could be disabled in config with e.g. `AGE_THRESHOLD: 0`.

Not a huge priority for me so no rush, and I'll likely get round to implementing this myself if you don't have time as IIRC I had some other tweaks I was planning to PR as well.

**Comment by monstermuffin** [comment]
I see. 

Would this be an option to use instead of space used or were you imagining both would be used in tandem?  

Also, if you are using 180GB of your cache which is 20%, is there any need for this to be moved to backing storage? 

Frigate in your use case is deleting the old files anyway (assuming by rolling you mean it can only use 180GB as a limit?) so moving stuff to backing doesn't sound like it would do much but be deleted when Frigate writes a new files to cache and deletes an old one to adhere to the storage limit.

**Comment by tigattack** [comment]
Yeah, the example I gave isn't a great example tbh as it doesn't really prove the need for this for the reasons you've stated.

I still think it would be a useful option to have, and my idea was it would work in tandem with the capacity limit, i.e. _if not at capacity limit, move on and check age limit if set_.

**Comment by macmacs** [comment]
Hi, this was somehow already tried out [here](https://github.com/jamesbradt/mergerfs-cache-mover/commit/3feb513e3de3fe77d13d6734654737c4aad5eea3).
I tried it out and it works pretty well.
Rebasing didn't yield and major conflicts -> DEV friendly :)

**Comment by monstermuffin** [comment]
> Hi, this was somehow already tried out [here](https://github.com/jamesbradt/mergerfs-cache-mover/commit/3feb513e3de3fe77d13d6734654737c4aad5eea3). I tried it out and it works pretty well. Rebasing didn't yield and major conflicts -> DEV friendly :)

Unless I'm being really stupid, the attached commit doesn't seem to implement time based moving, only an optimisation of gather files? 

There is a somewhat related discussion to the linked commit here: https://github.com/monstermuffin/mergerfs-cache-mover/issues/7

**Comment by macmacs** [comment]
You're right. Does not change too much from the existing logic.

**Comment by ulic75** [comment]
I would be interested in a feature like this.  Say for example I add 10 "linux-distro.iso" (as you like to say) every week and each iso is 5GB, for 50GB on a 1TB cache drive, the mover will never move them unless I set the threshold very low.  In the immediate this might not be a big deal, but I'd like these get to moved sooner rather than later so they get covered by the snapraid parity.

*note: I haven't actually implemented snapraid, mergerfs or the mover yet, I'm still in the planning stages in moving from unraid.

**Comment by monstermuffin** [comment]
Looking into this now guys. Did you foresee the calculation being done on creation time (ctime) or lastmod (mtime)?

I can probably support both but it would probably overcomplicate things. mtime probably makes more sense as if it's been accessed recently, likely should stay but this raises more questions:

If many files have an mtime that is not before the allowed time to move, but threshold has been met, does it just fail? Cache would fill up this way. It could fallback to usage but that's another can of worms.

**Comment by ulic75** [comment]
I think modified time is probably best.

I haven't got all my details worked out, but I think I might end up with a hybrid approach, where 6 days a week I run using percentage threshold and then once a week do a full cache clear.

Off topic slightly, what was your motivation to do a threshold clear and not to just do a full cache clear daily or however often the script was run?

**Comment by GlitchHound** [comment]
@ulic75 I'm not the author but in my case, my cache drive is plenty large enough, so by keeping the latest linux-distro.iso releases on cache, I get both (simplistic) read cache and proper write cache out of it. Generally speaking, the newest files are also a bit more likely to be accessed, so it's not a terrible heuristic for a read cache.

**Comment by crypticvoyage** [comment]
> Looking into this now guys. Did you foresee the calculation being done on creation time (ctime) or lastmod (mtime)?
> 
> I can probably support both but it would probably overcomplicate things. mtime probably makes more sense as if it's been accessed recently, likely should stay but this raises more questions:
> 
> If many files have an mtime that is not before the allowed time to move, but threshold has been met, does it just fail? Cache would fill up this way. It could fallback to usage but that's another can of worms.

Hi @monstermuffin, was this ever implemented? I looked at the config but didn't note anything related to age/time.
